File: packages/mcp/src/signing.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Ed25519 signing and verification for atrib records (§1.4).
 * Uses @noble/ed25519. pure JS, no native deps, audited.
 */

import * as ed from '@noble/ed25519'
import { sha512 } from '@noble/hashes/sha2.js'
import { base64urlEncode, base64urlDecode } from './base64url.js'
import { canonicalCrossAttestationInput, canonicalSigningInput } from './canon.js'
import { isValidEventTypeUri } from './types.js'
import type { AtribRecord, SignerEntry } from './types.js'

// @noble/ed25519 v3 needs sha512 wired via the hashes object
ed.hashes.sha512 = sha512
ed.hashes.sha512Async = (m) => Promise.resolve(sha512(m))

/**
 * Get the Ed25519 public key for a 32-byte private key.
 */
export async function getPublicKey(privateKey: Uint8Array): Promise<Uint8Array> {
  return ed.getPublicKeyAsync(privateKey)
}

/**
 * Sign an attribution record (§1.4.2).
 *
 * Step 1: Construct record with all fields except signature.
 * Step 2: Remove signature, apply JCS → signing input bytes.
 * Step 3: Ed25519 sign (RFC 8032 §5.1.6, Pure EdDSA).
 * Step 4: base64url encode the 64-byte signature.
 */
export async function signRecord(
  record: AtribRecord,
  privateKey: Uint8Array,
): Promise<AtribRecord> {
  const signingInput = canonicalSigningInput(record)
  const sigBytes = await ed.signAsync(signingInput, privateKey)
  const sigEncoded = base64urlEncode(sigBytes)
  if ('session_token' in record && record.session_token !== undefined) {
    return { ...record, signature: sigEncoded }
  }
  const { session_token: _, ...rest } = record as AtribRecord & { session_token?: never }
  return { ...rest, signature: sigEncoded } as AtribRecord
}

/**
 * Sign a transaction record using the §1.7.6 cross-attestation bytes.
 *
 * The returned record carries `signers[]` with this creator's signer entry
 * first, followed by any caller-supplied counterparty entries that already
 * signed the same canonical bytes. AP2 receipt JWTs are not valid inputs here:
 * they prove receipt acceptance, but they do not sign the atrib record bytes.
 */
export async function signTransactionRecord(
  record: AtribRecord,
  privateKey: Uint8Array,
  counterpartySigners: SignerEntry[] = [],
): Promise<AtribRecord> {
  const publicKey = await getPublicKey(privateKey)
  const creatorKey = base64urlEncode(publicKey)
  const transactionRecord = {
    ...record,
    creator_key: creatorKey,
    signature: '',
    signers: [],
  } as AtribRecord
  const signingInput = canonicalCrossAttestationInput(transactionRecord)
  const sigBytes = await ed.signAsync(signingInput, privateKey)
  const signer: SignerEntry = {
    creator_key: creatorKey,
    signature: base64urlEncode(sigBytes),
  }
  const signed = { ...transactionRecord, signers: [signer, ...counterpartySigners] }
  if ('session_token' in signed && signed.session_token !== undefined) {
    return signed
  }
  const { session_token: _, ...rest } = signed as AtribRecord & { session_token?: never }
  return rest as AtribRecord
}

/**
 * Verify an attribution record (§1.4.3). All 8 steps.
 * Returns true if and only if all steps pass.
 */
export async function verifyRecord(record: AtribRecord): Promise<boolean> {
  try {
    // Step 1: Decode creator_key → 32-byte public key
    const pubKeyBytes = base64urlDecode(record.creator_key)
    if (pubKeyBytes.length !== 32) return false

    const isTransaction = record.event_type === 'https://atrib.dev/v1/types/transaction'
    const hasSignersArray = Array.isArray(record.signers) && record.signers.length > 0

    if (isTransaction && hasSignersArray) {
      const creatorSigner = record.signers!.find(
        (entry) => entry.creator_key === record.creator_key,
      )
      if (!creatorSigner) return false
      const sigBytes = base64urlDecode(creatorSigner.signature)
      if (sigBytes.length !== 64) return false
      const verifyInput = canonicalCrossAttestationInput(record)
      const valid = await ed.verifyAsync(sigBytes, verifyInput, pubKeyBytes)
      if (!valid) return false
    } else {
      // Step 2: Decode signature → 64-byte signature
      const sigBytes = base64urlDecode(record.signature)
      if (sigBytes.length !== 64) return false

      // Step 3: Remove signature, apply JCS → verification input bytes
      const verifyInput = canonicalSigningInput(record)

      // Step 4: Verify Ed25519 signature (RFC 8032 §5.1.7)
      const valid = await ed.verifyAsync(sigBytes, verifyInput, pubKeyBytes)
      if (!valid) return false
    }

    // Step 5: spec_version must be "atrib/1.0"
    if (record.spec_version !== 'atrib/1.0') return false

    // Step 6: event_type must be a syntactically-valid absolute URI per spec 1.4.5.
    // Recognition (whether the URI is in atrib normative set) is informational
    // and does NOT gate verification; an extension URI passes this check.
    if (!isValidEventTypeUri(record.event_type)) return false

    // Step 7: timestamp must be a finite number and not >5 min in the future.
    // NaN, Infinity, negative values are all invalid.
    // The log server (§2.6.1 Step 4) uses a wider 10-minute window to
    // account for network/queue delay between signing and log submission.
    if (!Number.isInteger(record.timestamp) || record.timestamp < 0) return false
    const fiveMinutesFromNow = Date.now() + 5 * 60 * 1000
    if (record.timestamp > fiveMinutesFromNow) return false

    // Step 8: context_id must be exactly 32 lowercase hex chars
    if (!/^[0-9a-f]{32}$/.test(record.context_id)) return false

    return true
  } catch {
    return false
  }
}

File: python/src/atrib/signing.py
# SPDX-License-Identifier: Apache-2.0
"""Ed25519 signing and verification for atrib records (§1.4).

Port of ``packages/mcp/src/signing.ts``. Keys are raw 32-byte Ed25519
seeds (§1.4.1) — never 64-byte NaCl expanded format. Signatures are Pure
EdDSA (RFC 8032), base64url no padding (86 chars).
"""

from __future__ import annotations

import time
from collections.abc import Mapping
from typing import cast

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from .canon import canonical_cross_attestation_input, canonical_signing_input
from .encoding import base64url_decode, base64url_encode
from .types import (
    EVENT_TYPE_TRANSACTION_URI,
    SPEC_VERSION,
    AtribRecord,
    SignerEntry,
    is_valid_event_type_uri,
)

_CONTEXT_ID_HEX = frozenset("0123456789abcdef")
_FIVE_MINUTES_MS = 5 * 60 * 1000


def get_public_key(private_key: bytes) -> bytes:
    """Raw 32-byte Ed25519 public key for a 32-byte seed."""
    if len(private_key) != 32:
        raise ValueError("atrib: private key must be a 32-byte Ed25519 seed")
    key = Ed25519PrivateKey.from_private_bytes(private_key)
    return key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)


def sign_record(record: Mapping[str, object], private_key: bytes) -> AtribRecord:
    """§1.4.2: remove ``signature``, JCS-serialize, Ed25519-sign, base64url.

    Returns a new record dict with ``signature`` set. Optional fields keep
    their exact presence/absence from the input (omission ≠ null)."""
    if len(private_key) != 32:
        raise ValueError("atrib: private key must be a 32-byte Ed25519 seed")
    signing_input = canonical_signing_input(record)
    key = Ed25519PrivateKey.from_private_bytes(private_key)
    signature = base64url_encode(key.sign(signing_input))
    signed = {k: v for k, v in record.items() if k != "signature"}
    signed["signature"] = signature
    return cast(AtribRecord, signed)


def sign_transaction_record(
    record: Mapping[str, object],
    private_key: bytes,
    counterparty_signers: list[SignerEntry] | None = None,
) -> AtribRecord:
    """§1.7.6: sign a transaction record's cross-attestation bytes. The
    returned record carries this creator's signer entry first, followed by
    caller-supplied counterparty entries over the same canonical bytes."""
    creator_key = base64url_encode(get_public_key(private_key))
    txn: dict[str, object] = {k: v for k, v in record.items() if k != "signers"}
    txn["creator_key"] = creator_key
    txn["signature"] = ""
    txn["signers"] = []
    signing_input = canonical_cross_attestation_input(txn)
    key = Ed25519PrivateKey.from_private_bytes(private_key)
    entry: SignerEntry = {
        "creator_key": creator_key,
        "signature": base64url_encode(key.sign(signing_input)),
    }
    txn["signers"] = [entry, *(counterparty_signers or [])]
    return cast(AtribRecord, txn)


def sign_transaction_attestation(
    record: Mapping[str, object], private_key: bytes
) -> SignerEntry:
    """Create one counterparty signer entry over an existing transaction
    record's §1.7.6 cross-attestation bytes."""
    if record.get("event_type") != EVENT_TYPE_TRANSACTION_URI:
        raise ValueError("atrib: transaction attestation requires a transaction record")
    key = Ed25519PrivateKey.from_private_bytes(private_key)
    signature = key.sign(canonical_cross_attestation_input(record))
    return {
        "creator_key": base64url_encode(get_public_key(private_key)),
        "signature": base64url_encode(signature),
    }


def _verify_ed25519(signature: bytes, message: bytes, public_key: bytes) -> bool:
    try:
        Ed25519PublicKey.from_public_bytes(public_key).verify(signature, message)
        return True
    except (InvalidSignature, ValueError):
        return False


def verify_record(record: Mapping[str, object], *, now_ms: int | None = None) -> bool:
    """§1.4.3 verification, all 8 steps. True iff every step passes.

    ``now_ms`` injects the clock for corpus tests; defaults to wall time.
    """
    try:
        # Step 1: creator_key decodes to 32 bytes.
        creator_key = record.get("creator_key")
        if not isinstance(creator_key, str):
            return False
        pub_key = base64url_decode(creator_key)
        if len(pub_key) != 32:
            return False

        is_transaction = record.get("event_type") == EVENT_TYPE_TRANSACTION_URI
        signers = record.get("signers")
        has_signers = isinstance(signers, list) and len(signers) > 0

        if is_transaction and has_signers:
            creator_signer = next(
                (
                    entry
                    for entry in cast(list[dict[str, object]], signers)
                    if isinstance(entry, dict) and entry.get("creator_key") == creator_key
                ),
                None,
            )
            if creator_signer is None:
                return False
            entry_sig = creator_signer.get("signature")
            if not isinstance(entry_sig, str):
                return False
            sig = base64url_decode(entry_sig)
            if len(sig) != 64:
                return False
            if not _verify_ed25519(sig, canonical_cross_attestation_input(record), pub_key):
                return False
        else:
            # Step 2: signature decodes to 64 bytes.
            signature = record.get("signature")
            if not isinstance(signature, str):
                return False
            sig = base64url_decode(signature)
            if len(sig) != 64:
                return False
            # Steps 3–4: JCS minus signature, Ed25519 verify.
            if not _verify_ed25519(sig, canonical_signing_input(record), pub_key):
                return False

        # Step 5: spec_version.
        if record.get("spec_version") != SPEC_VERSION:
            return False

        # Step 6: event_type is a syntactically-valid absolute URI.
        if not is_valid_event_type_uri(record.get("event_type")):
            return False

        # Step 7: timestamp is a non-negative integer, not >5min future.
        timestamp = record.get("timestamp")
        if not isinstance(timestamp, int) or isinstance(timestamp, bool) or timestamp < 0:
            return False
        now = int(time.time() * 1000) if now_ms is None else now_ms
        if timestamp > now + _FIVE_MINUTES_MS:
            return False

        # Step 8: context_id is exactly 32 lowercase hex chars.
        context_id = record.get("context_id")
        if (
            not isinstance(context_id, str)
            or len(context_id) != 32
            or not all(c in _CONTEXT_ID_HEX for c in context_id)
        ):
            return False

        return True
    except (ValueError, TypeError):
        return False
